File: 05-seguranca-ctf/backend/php/src/Controllers/AuthController.php
<?php

declare(strict_types=1);

namespace App\Controllers;

use App\Database;
use App\JwtManager;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class AuthController
{
    public function login(Request $request, Response $response): Response
    {
        $body = $request->getParsedBody();
        if ($body === null) {
            $raw = (string) $request->getBody();
            $body = json_decode($raw, true);
        }
        if (!is_array($body)) {
            Database::logRequest([
                'method' => 'POST', 'path' => '/api/auth/login', 'queryParams' => '',
                'body' => '', 'statusCode' => 400, 'sqlQuery' => '',
                'responsePreview' => 'Invalid JSON',
                'ip' => self::getIp($request),
            ]);
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'JSON invalido',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $email    = $body['email'] ?? '';
        $password = $body['password'] ?? '';

        $sqlQuery = "SELECT * FROM users WHERE email = '$email' AND password = '$password'";

        $db = Database::getInstance();
        try {
            $sql = "SELECT * FROM users WHERE email = :email AND password = '$password'";
            $stmt = $db->prepare($sql);
            $stmt->execute([':email' => $email]);
            $user = $stmt->fetch();
        } catch (\Exception $e) {
            Database::logRequest([
                'method' => 'POST', 'path' => '/api/auth/login', 'queryParams' => '',
                'body' => json_encode($body), 'statusCode' => 500,
                'sqlQuery' => $sqlQuery, 'responsePreview' => $e->getMessage(),
                'ip' => self::getIp($request),
            ]);
            $response->getBody()->write(json_encode([
                'error' => 'Database error',
                'message' => $e->getMessage(),
            ]));
            return $response->withStatus(500)->withHeader('Content-Type', 'application/json');
        }

        if (!$user) {
            Database::logRequest([
                'method' => 'POST', 'path' => '/api/auth/login', 'queryParams' => '',
                'body' => json_encode($body), 'statusCode' => 401,
                'sqlQuery' => $sqlQuery, 'responsePreview' => 'Invalid credentials',
                'ip' => self::getIp($request),
            ]);
            $response->getBody()->write(json_encode([
                'error' => 'Unauthorized',
                'message' => 'Credenciais invalidas',
            ]));
            return $response->withStatus(401)->withHeader('Content-Type', 'application/json');
        }

        $token = JwtManager::generate([
            'sub'   => $user['id'],
            'email' => $user['email'],
            'role'  => $user['role'],
        ]);

        Database::logRequest([
            'method' => 'POST', 'path' => '/api/auth/login', 'queryParams' => '',
            'body' => json_encode($body), 'statusCode' => 200,
            'sqlQuery' => $sqlQuery,
            'responsePreview' => json_encode(['id' => $user['id'], 'email' => $user['email'], 'role' => $user['role']]),
            'ip' => self::getIp($request),
        ]);

        $response->getBody()->write(json_encode([
            'token' => $token,
            'user'  => [
                'id'    => $user['id'],
                'nome'  => $user['nome'],
                'email' => $user['email'],
                'role'  => $user['role'],
            ],
        ]));
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function register(Request $request, Response $response): Response
    {
        $body = $request->getParsedBody();
        if ($body === null) {
            $raw = (string) $request->getBody();
            $body = json_decode($raw, true);
        }
        if (!is_array($body)) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'JSON invalido',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $nome     = $body['nome'] ?? '';
        $email    = $body['email'] ?? '';
        $password = $body['password'] ?? '';
        $role     = $body['role'] ?? 'user';

        if (!$nome || !$email || !$password) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'Campos obrigatorios: nome, email, password',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $sqlQuery = "INSERT INTO users (nome, email, password, role) VALUES ('$nome', '$email', '$password', '$role')";

        $db = Database::getInstance();
        try {
            $stmt = $db->prepare("INSERT INTO users (nome, email, password, role) VALUES (?, ?, ?, ?)");
            $stmt->execute([$nome, $email, $password, $role]);
        } catch (\Exception $e) {
            Database::logRequest([
                'method' => 'POST', 'path' => '/api/auth/register', 'queryParams' => '',
                'body' => json_encode($body), 'statusCode' => 409,
                'sqlQuery' => $sqlQuery, 'responsePreview' => $e->getMessage(),
                'ip' => self::getIp($request),
            ]);
            $response->getBody()->write(json_encode([
                'error' => 'Conflict',
                'message' => 'Email ja cadastrado',
            ]));
            return $response->withStatus(409)->withHeader('Content-Type', 'application/json');
        }

        $stmt = $db->prepare("SELECT id, nome, email, role FROM users WHERE email = ?");
        $stmt->execute([$email]);
        $newUser = $stmt->fetch();

        Database::logRequest([
            'method' => 'POST', 'path' => '/api/auth/register', 'queryParams' => '',
            'body' => json_encode($body), 'statusCode' => 201,
            'sqlQuery' => $sqlQuery, 'responsePreview' => json_encode($newUser),
            'ip' => self::getIp($request),
        ]);

        $response->getBody()->write(json_encode([
            'message' => 'Usuario criado com sucesso',
            'user' => $newUser,
        ]));
        return $response->withStatus(201)->withHeader('Content-Type', 'application/json');
    }

    public function forgotPassword(Request $request, Response $response): Response
    {
        $body = $request->getParsedBody();
        if ($body === null) {
            $raw = (string) $request->getBody();
            $body = json_decode($raw, true);
        }
        if (!is_array($body)) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'JSON invalido',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $email = $body['email'] ?? '';

        if (!$email) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => "Campo 'email' obrigatorio",
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        try {
            $smtpConfigPath = dirname(__DIR__, 2) . '/config/smtp.json';
            $raw = file_get_contents($smtpConfigPath);
            if ($raw === false) {
                throw new \RuntimeException("Cannot read SMTP configuration from $smtpConfigPath");
            }
            $smtpConfig = json_decode($raw, true);

            $smtpConnection = $smtpConfig['host'] . ':' . $smtpConfig['port'];

            $db = Database::getInstance();
            $stmt = $db->prepare("SELECT id, nome, email FROM users WHERE email = ?");
            $stmt->execute([$email]);
            $user = $stmt->fetch();

            if ($user) {
                $resetToken = JwtManager::generate([
                    'sub'   => $user['id'],
                    'email' => $user['email'],
                    'role'  => 'reset',
                ]);
                $resetUrl = "http://localhost:5173/reset-password?token=$resetToken";
                @fwrite(fopen('php://stderr', 'w'), "[SMTP] Sending reset email to {$user['email']} via $smtpConnection\n");
            }

            Database::logRequest([
                'method' => 'POST', 'path' => '/api/auth/forgot-password', 'queryParams' => '',
                'body' => json_encode(['email' => $email]), 'statusCode' => 200,
                'sqlQuery' => "SELECT id, nome, email FROM users WHERE email = '$email'",
                'responsePreview' => 'Reset email sent',
                'ip' => self::getIp($request),
            ]);

            $response->getBody()->write(json_encode([
                'message' => 'Se o email estiver cadastrado, voce recebera as instrucoes de recuperacao.',
            ]));
            return $response->withHeader('Content-Type', 'application/json');
        } catch (\Throwable $e) {
            Database::logRequest([
                'method' => 'POST', 'path' => '/api/auth/forgot-password', 'queryParams' => '',
                'body' => json_encode(['email' => $email]), 'statusCode' => 500,
                'sqlQuery' => '', 'responsePreview' => $e->getMessage(),
                'ip' => self::getIp($request),
            ]);

            $response->getBody()->write(json_encode([
                'error'   => 'Internal Server Error',
                'message' => $e->getMessage(),
                'stack'   => array_slice(explode("\n", $e->getTraceAsString()), 0, 8),
            ]));
            return $response->withStatus(500)->withHeader('Content-Type', 'application/json');
        }
    }

    private static function getIp(Request $request): string
    {
        return $request->getHeaderLine('X-Forwarded-For') ?: ($request->getServerParams()['REMOTE_ADDR'] ?? 'unknown');
    }
}

File: 05-seguranca-ctf/backend/java/src/main/java/com/treinamento/ctf/controller/AuthController.java
package com.treinamento.ctf.controller;

import com.treinamento.ctf.middleware.JwtUtil;
import com.treinamento.ctf.middleware.RequestLogFilter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.*;

import java.io.FileInputStream;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private JwtUtil jwtUtil;

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> body) {
        String email    = body.getOrDefault("email", "");
        String password = body.getOrDefault("password", "");

        String sql = "SELECT * FROM users WHERE email = ? AND password = '" + password + "'";

        RequestLogFilter.CURRENT_SQL.set(
                "SELECT * FROM users WHERE email = '" + email + "' AND password = '" + password + "'");

        try {
            List<Map<String, Object>> rows = jdbcTemplate.queryForList(sql, email);

            if (rows.isEmpty()) {
                Map<String, Object> err = new LinkedHashMap<>();
                err.put("error", "Unauthorized");
                err.put("message", "Credenciais invalidas");
                return ResponseEntity.status(401).body(err);
            }

            Map<String, Object> user = rows.get(0);

            Integer userId = (Integer) user.get("ID");
            String  uEmail = (String)  user.get("EMAIL");
            String  role   = (String)  user.get("ROLE");
            String  nome   = (String)  user.get("NOME");

            String token = jwtUtil.generateToken(userId, uEmail, role);

            Map<String, Object> userMap = new LinkedHashMap<>();
            userMap.put("id", userId);
            userMap.put("nome", nome);
            userMap.put("email", uEmail);
            userMap.put("role", role);

            Map<String, Object> result = new LinkedHashMap<>();
            result.put("token", token);
            result.put("user", userMap);

            return ResponseEntity.ok(result);

        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Database error");
            err.put("message", ex.getMessage());
            return ResponseEntity.status(500).body(err);
        }
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody Map<String, Object> body) {
        String nome     = (String) body.getOrDefault("nome", "");
        String email    = (String) body.getOrDefault("email", "");
        String password = (String) body.getOrDefault("password", "");
        String role     = (String) body.getOrDefault("role", "user");

        if (nome.isEmpty() || email.isEmpty() || password.isEmpty()) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Bad Request");
            err.put("message", "Campos obrigatorios: nome, email, password");
            return ResponseEntity.status(400).body(err);
        }

        String sqlQuery = "INSERT INTO users (nome, email, password, role) VALUES ('" +
                nome + "', '" + email + "', '" + password + "', '" + role + "')";
        RequestLogFilter.CURRENT_SQL.set(sqlQuery);

        try {
            jdbcTemplate.update(
                "INSERT INTO users (nome, email, password, role) VALUES (?, ?, ?, ?)",
                nome, email, password, role
            );
        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Conflict");
            err.put("message", "Email ja cadastrado");
            return ResponseEntity.status(409).body(err);
        }

        try {
            Map<String, Object> newUser = jdbcTemplate.queryForMap(
                "SELECT id, nome, email, role FROM users WHERE email = ?", email
            );

            Map<String, Object> userMap = new LinkedHashMap<>();
            userMap.put("id", newUser.get("ID"));
            userMap.put("nome", newUser.get("NOME"));
            userMap.put("email", newUser.get("EMAIL"));
            userMap.put("role", newUser.get("ROLE"));

            Map<String, Object> result = new LinkedHashMap<>();
            result.put("message", "Usuario criado com sucesso");
            result.put("user", userMap);

            return ResponseEntity.status(201).body(result);

        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Database error");
            err.put("message", ex.getMessage());
            return ResponseEntity.status(500).body(err);
        }
    }

    @PostMapping("/forgot-password")
    public ResponseEntity<?> forgotPassword(@RequestBody Map<String, String> body) {
        String email = body.getOrDefault("email", "");

        if (email.isEmpty()) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Bad Request");
            err.put("message", "Campo 'email' obrigatorio");
            return ResponseEntity.status(400).body(err);
        }

        try {
            // Load SMTP configuration from config file
            String configPath = System.getProperty("user.dir") + "/config/smtp.json";
            FileInputStream fis = new FileInputStream(configPath);
            byte[] data = fis.readAllBytes();
            fis.close();
            String smtpConfig = new String(data);

            // Parse and use SMTP config to send reset email
            Map<String, Object> user = jdbcTemplate.queryForMap(
                "SELECT id, nome, email FROM users WHERE email = ?", email
            );

            if (user != null) {
                String resetToken = jwtUtil.generateToken(
                    (Integer) user.get("ID"),
                    (String) user.get("EMAIL"),
                    "reset"
                );
                System.out.println("[SMTP] Sending reset email to " + user.get("EMAIL") +
                    " via " + smtpConfig);
            }

            Map<String, Object> result = new LinkedHashMap<>();
            result.put("message", "Se o email estiver cadastrado, voce recebera as instrucoes de recuperacao.");
            return ResponseEntity.ok(result);

        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Internal Server Error");
            err.put("message", ex.getMessage());

            String[] stackLines = new String[0];
            if (ex.getStackTrace() != null) {
                stackLines = new String[Math.min(8, ex.getStackTrace().length)];
                for (int i = 0; i < stackLines.length; i++) {
                    stackLines[i] = ex.getStackTrace()[i].toString();
                }
            }
            err.put("stack", java.util.Arrays.asList(stackLines));

            return ResponseEntity.status(500).body(err);
        }
    }
}
